ouroboros-network/test/Test/Ouroboros/Network/PeerSelection.hs

```diff
@@ -156,16 +154,151 @@ isEmptyEnv GovernorMockEnvironment {
+-- cycle of continuous activity. Due to the way the I\/O simulator manages
+-- virtual time, these bugs exhibits themselves by infinite trace activity
+-- without time advancing.
+--
+-- It is important to catch these bugs early in the set of tests, since it is
+-- hard to write many of the other more interesting properties if there are
+-- these kinds of livelock bugs. Or to put it another way, the other properties
+-- can be expressed more simply if they can assume within event traces that the
+-- time always advances after some finite number of events.
+--
+prop_governor_nolivelock :: GovernorMockEnvironment -> Property
+prop_governor_nolivelock env =
+    let trace = take 500 .
+                selectGovernorEvents .
+                selectPeerSelectionTraceEvents $
+                  runGovernorInMockEnvironment env
+     in case tooManyEventsBeforeTimeAdvances 1000 trace of
+          Nothing -> property True
+          Just (t, es) ->
+            counterexample
+              ("over 1000 events at time: " ++ show t ++ "\n" ++
+               "first 50 events: " ++ (unlines . map show . take 50 $ es)) $
+            property True
+
+
+-- | Scan the trace and return any occurrence where we have at least threshold
+-- events before virtual time moves on. Return the tail of the trace from that
+-- point on.
+--
+tooManyEventsBeforeTimeAdvances :: Int -> [(Time, e)] -> Maybe (Time, [e])
+tooManyEventsBeforeTimeAdvances _         []     = Nothing
+tooManyEventsBeforeTimeAdvances threshold trace0 =
+    go [ (t, diffTime t' t, e)
+       | ((t, e), (t', _)) <- zip trace0 (tail trace0) ]
   where
-    go (a:as) (b:bs)
-      | diffTime b a < adequate = False
-      | otherwise               = go as bs
-    go _ _ = True
+    go []                = Nothing
+    go trace@((t,_,_):_) = case countdown threshold trace of
+                             Just es' -> go es'
+                             Nothing  -> Just (t, trace')
+                               where
+                                 trace' = take threshold [ e | (_,_,e) <- trace ]
+
+    countdown 0 _  = Nothing
+    countdown _ [] = Just []
+    countdown n ((_t,dt,_e):es)
+      | dt == 0    = countdown (n-1) es
+      | otherwise  = Just es
+
+
+-- | A coverage property, much like 'prop_governor_nofail' but we look to see
+-- that we get adequate coverage of the state space. We look for all the trace
+-- events that the governor can produce, and tabules which ones we see.
+--
+prop_governor_trace_coverage :: GovernorMockEnvironment -> Property
+prop_governor_trace_coverage env =
+    let trace = take 5000 .
+                selectPeerSelectionTraceEvents $
+                  runGovernorInMockEnvironment env
+
+        traceNumsSeen  = collectTraces trace
+        traceNamesSeen = allTraceNames `Map.restrictKeys` traceNumsSeen
+
+     in coverTable "trace events" [ (n, 1) | n <- Map.elems allTraceNames ] $
+        tabulate   "trace events" (Map.elems traceNamesSeen)
+        True
```

Comment:
I think `tabulate` only reports, so this property will never fail.

Reply:
Yes this is not intended to fail automatically yet. There are indeed a few we do not cover, this test reveals. There's no point failing until we extend the coverage.